src/go/k8s/internal/client/factory.go

```diff
@@ -0,0 +1,183 @@
+package client
+
+import (
+	"context"
+	"errors"
+
+	"github.com/go-logr/logr"
+	"github.com/redpanda-data/common-go/rpadmin"
+	"github.com/redpanda-data/helm-charts/pkg/redpanda"
+	redpandav1alpha1 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha1"
+	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/src/go/k8s/api/redpanda/v1alpha2"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	ErrEmptyBrokerList             = errors.New("empty broker list")
+	ErrEmptyURLList                = errors.New("empty url list")
+	ErrInvalidKafkaClientObject    = errors.New("cannot initialize Kafka API client from given object")
+	ErrInvalidRedpandaClientObject = errors.New("cannot initialize Redpanda Admin API client from given object")
+	ErrUnsupportedSASLMechanism    = errors.New("unsupported SASL mechanism")
+)
+
+// ClientFactory is responsible for creating both high-level and low-level clients used in our
+// controllers.
+//
+// Calling its `Kafka*` methods will initialize a low-level kgo.Client instance
```

Comment:
Not a convention just a fun fact, if you enclose `kgo.Client` in `[]`'s godoc will automatically link it to the documentation of `kgo.Client`.